src/utils/few_modes_ft.cpp

```diff
@@ -83,8 +83,118 @@ FewModesFT::FewModesFT(parthenon::ParameterInput *pin, parthenon::StateDescripto
       "random_num", 3, num_modes, 2);
   random_num_host_ = Kokkos::create_mirror_view(random_num_);
 
-  rng_.seed(rseed);
-  dist_ = std::uniform_real_distribution<>(-1.0, 1.0);
+  bool is_restart = pin->DoesParameterExist("few_modes_ft", "var_hat_0_0_r");
+  if (is_restart) {
+    // Restore acceleration field in Fourier space
+    std::cout << "restoring acceleration field...\n\n";
+
+    // read var_hat
+    std::cout << "reading var_hat...\n";
+    auto var_hat_host = Kokkos::create_mirror_view(var_hat_);
+    for (int i = 0; i < 3; i++) {
+      for (int m = 0; m < num_modes; m++) {
+        auto real = pin->GetReal("few_modes_ft", "var_hat_" + std::to_string(i) + "_" +
+                                                     std::to_string(m) + "_r");
+        auto imag = pin->GetReal("few_modes_ft", "var_hat_" + std::to_string(i) + "_" +
+                                                     std::to_string(m) + "_i");
+        std::cout << "(" << i << "," << m << "): " << real << "\t" << imag << "\n";
+        var_hat_host(i, m) = Complex(real, imag);
+      }
+    }
+    Kokkos::deep_copy(var_hat_, var_hat_host);
+
+    // read var_hat_new
+    std::cout << "reading var_hat_new...\n";
+    auto var_hat_new_host = Kokkos::create_mirror_view(var_hat_new_);
+    for (int i = 0; i < 3; i++) {
+      for (int m = 0; m < num_modes; m++) {
+        auto real_new = pin->GetReal("few_modes_ft", "var_hat_new_" + std::to_string(i) +
+                                                         "_" + std::to_string(m) + "_r");
+        auto imag_new = pin->GetReal("few_modes_ft", "var_hat_new_" + std::to_string(i) +
+                                                         "_" + std::to_string(m) + "_i");
+        std::cout << "(" << i << "," << m << "): " << real_new << "\t" << imag_new
+                  << "\n";
+        var_hat_new_host(i, m) = Complex(real_new, imag_new);
+      }
+    }
+    Kokkos::deep_copy(var_hat_new_, var_hat_new_host);
+
+    // Restore rng state
+    {
+      std::istringstream iss(pin->GetString("few_modes_ft", "state_rng"));
+      iss >> rng_;
+      std::cout << "rng state: " << rng_ << "\n";
+    }
+
+    // Restore dist
+    {
+      std::istringstream iss(pin->GetString("few_modes_ft", "state_dist"));
+      iss >> dist_;
+      std::cout << "dist state: " << dist_ << "\n";
+    }
+  } else {
+    // this is NOT a restart
+    rng_.seed(rseed);
+    dist_ = std::uniform_real_distribution<>(-1.0, 1.0);
+  }
+}
+
+void FewModesFT::SaveStateBeforeOutput(Mesh *mesh, ParameterInput *pin) {
+  // save state before output
+  auto md = mesh->mesh_data.Get();
+  auto pmb = md->GetBlockData(0)->GetBlockPointer();
+  auto hydro_pkg = pmb->packages.Get("Hydro");
+  auto num_modes = pin->GetInteger("precipitator/driving", "num_modes");
+
+  // Save acceleration field in Fourier space
+  auto var_hat_host = Kokkos::create_mirror_view(var_hat_);
```

Comment:
```suggestion
  auto var_hat_host = Kokkos::create_mirror_view_and_copy(var_hat_);
```
Without copy anything could be written to disk.

Reply:
well this might explain some things. what does `create_mirror_view` do vs `create_mirror_view_and_copy`?

Reply:
ah nvm. i missed the asymmetry here: https://github.com/parthenon-hpc-lab/athenapk/blob/f334d206f78ee2e8997f0c93390e361da99ed4f1/src/pgen/turbulence.cpp#L456

Reply:
> well this might explain some things. what does `create_mirror_view` do vs `create_mirror_view_and_copy`?

create -> malloc in the correct memory space with metadata information (layout) being identical to the original view
create_and_copy -> same as create but with a following deep_copy from the source